### ISSUE TITLE
Image component and optional context values

### ISSUE THREAD

**Issue by onepayclick** [vapor-community/HTMLKit#104]
Hi, love the library, but I'm having an extremely hard time trying to display an image using an url that comes from an optional field from the model
These are some scenarios that I'm facing
![image](https://user-images.githubusercontent.com/64233519/200324896-d33db802-6e26-45fa-b847-81bde4f6a57b.png)

I had to create a computed var to try to drill down the model and fetch the url or default to "" and still doesn't work
```
var profile_thumb_url: String {
    attributes.profile_thumb.data?.attributes.url ?? ""
}
```
![image](https://user-images.githubusercontent.com/64233519/200325077-8444c830-3598-497d-8a5a-488d5d5b5779.png)

If I try to use the whole path I get these scenarios
![image](https://user-images.githubusercontent.com/64233519/200326273-948aa975-cbed-42be-9ecd-3d5f42d21b90.png)

And if I try to get fancy 
![image](https://user-images.githubusercontent.com/64233519/200328005-d6b8ecb0-6aeb-49a0-ac03-80910b4cfb8c.png)

**Comment by mattesmohr** [comment]
Sorry, to hear.  I hope, I can help.

Do you want to use the component 'image' or the html element 'image'? 

and

Do you use TemplateValue or without?

**Comment by onepayclick** [comment]
Hi :)

yes I'm using a TemplateValue. 
I want to use the HTML element image (<img>)

I just noticed that *HTLMKitComponents.Image(source:)* its just a wrapper around *HTMLKit.Image().source(TemplateValue.rawValue)*  and I tried both but I'm not sure why the src property is empty when displaying the page :(

**Comment by onepayclick** [comment]
Just an update

I tried using a normal type instead of a TemplateValue<T> just to be sure the issue was not at my side and the model was correctly populated
![image](https://user-images.githubusercontent.com/64233519/200548771-aa38489f-0a72-49f2-aae0-1d34997cc797.png)

and I was able to present the image using 
Image().source(profiles[0].attributes.profile_thumb.data.attributes.url)
I had to force *profiles[0]* because I can't use normal trays with ForEach, they need to be TemplateValues.

The problem could be somewhere in the unwrapping of .dynamic maybe

**Comment by onepayclick** [comment]
Btw, not sure it makes a difference here but I'm using Linux

**Comment by onepayclick** [comment]
I'm using this hack to repackage the value until we can find a solution 

```
func makeConstant<Value>(_ value: Value) -> TemplateValue<Value> {
    return .constant(value)
}
```

but then I found another issue. The HTMLKitComponents Image type doesn't seem to have access to any of the HTMLKit methods like .style, .class, .height

![image](https://user-images.githubusercontent.com/64233519/200560297-43dbe308-2bd4-4a65-8763-52cb65bb35f5.png)

This seems to be the case with other components too

**Comment by mattesmohr** [comment]
I am currently at work. I will look into it tonight, when I am at home.

**Comment by mattesmohr** [comment]
> Hi :)
> 
> yes I'm using a TemplateValue. I want to use the HTML element image "<img src..."
> 
> I just noticed that _HTLMKitComponents.Image(source:)_ its just a wrapper around _HTMLKit.Image().source(TemplateValue.rawValue)_ and I tried both but I'm not sure why the src property is empty when displaying the page :(

If you want to use the HTML element 'image', then you should use just Image(). The component 'image' comes from the UI component library I am working on. That you able to use it, tells me, that you running against the main branch.  If you do so, be aware that you might use implementations, wich could be work in progress and not meant for release yet.

I am currently working on an update of the documentation. The documentation will explain and extend the thought of the UI component library.

**Comment by mattesmohr** [comment]
> Just an update
> 
> I tried using a normal type instead of a TemplateValue just to be sure the issue was not at my side and the model was correctly populated ![image](https://user-images.githubusercontent.com/64233519/200548771-aa38489f-0a72-49f2-aae0-1d34997cc797.png)
> 
> and I was able to present the image using `Image().source(profiles[0].attributes.profile_thumb.data.attributes.url)` I had to force _profiles[0]_ because I can't use normal arrays with ForEach, they need to be TemplateValues.
> 
> The problem could be somewhere in the unwrapping of .dynamic maybe
> 
> [Update]: Image(source:) body uses rawValue ![image](https://user-images.githubusercontent.com/64233519/200553350-369eb06b-eb1a-4ba4-9d3c-4e79ea192561.png)
> 
> but TemplateValue rawValue returns an empty string if the value is .dynamic ![image](https://user-images.githubusercontent.com/64233519/200553479-ca627199-fef2-4a26-9145-f56a1c2a53a9.png)
> 
> if you render the page using this function ![image](https://user-images.githubusercontent.com/64233519/200553784-da439183-c42c-46e4-ad62-396279271595.png) it seems to package your value into the _.dynamic_ case instead of _.constant_
> 
> I tried unpacking the .dynamic but it seems that it wraps the value in an HTMLContext and for the looks of it, and HTMLContext does not stores the value, but a key path to it

Yup, you are on the right track.  TemplateValue wraps the value behind a KeyPath. Thats why you can't use string interpolation for example. The .rawValue was more meant as an short-term hack, not really as long-term solution. 

TemplateValue and the context handling need some love and I am already on it.

**Comment by mattesmohr** [comment]
> 

Makes no difference :-)

**Comment by mattesmohr** [comment]
> I'm using this hack to repackage the value until we can find a solution
> 
> ```
> func makeConstant<Value>(_ value: Value) -> TemplateValue<Value> {
>     return .constant(value)
> }
> ```
> 
> but then I found another issue. The HTMLKitComponents Image type doesn't seem to have access to any of the HTMLKit methods like .style, .class, .height
> 
> ![image](https://user-images.githubusercontent.com/64233519/200560297-43dbe308-2bd4-4a65-8763-52cb65bb35f5.png)
> 
> This seems to be the case with other components too

All the UI components come with a predefined CSS. Overriding the style is a discussion worth. I haven thing about it yet. So if you the components library, to you use the CSS file too?

**Comment by onepayclick** [comment]
> > Hi :)
> > yes I'm using a TemplateValue. I want to use the HTML element image "<img src..."
> > I just noticed that _HTLMKitComponents.Image(source:)_ its just a wrapper around _HTMLKit.Image().source(TemplateValue.rawValue)_ and I tried both but I'm not sure why the src property is empty when displaying the page :(
> 
> If you want to use the HTML element 'image', then you should use just Image(). The component 'image' comes from the UI component library I am working on. That you able to use it, tells me, that you running against the main branch. If you do so, be aware that you might use implementations, wich could be work in progress and not meant for release yet.
> 
> I am currently working on an update of the documentation. The documentation will explain and extend the thought of the UI component library.

Oh I'm using the deprecated Components repo, I saw that you moved it inside the HTMLKit library but I wanted to avoid targeting 'main' so I can keep a stable version.

> All the UI components come with a predefined CSS. Overriding the style is a discussion worth. I haven't thought about it. So if you use the components library, do you use the CSS file too? How did you get to manage it? I haven't released instructions yet :-)

Yes I'm using the css files :) and I agree that having a fixed style instead of being able to call '.style' would be closer to SwiftUI and preferable. 
I used your example website to understand how you set it all up and did the same 😁

**Comment by mattesmohr** [comment]
Ah, that's sharp! 

Didn't suspect someone will use the component library. But am happy, someone do. If you miss something (similar or in comparison to SwiftUI), feel free to open an issue.

**Comment by mattesmohr** [comment]
Ups, sorry. Wrong button.

**Comment by mattesmohr** [comment]
I want to let you know, that we remove the context management. You won't have this issue anymore. You can read more about in  #118. You can already use the 3.0.0-alpha.1 release, if you like. But keep in mind its still alpha.
